api/src/passports/single-use-code.strategy.ts

```diff
@@ -0,0 +1,199 @@
+import { Strategy } from 'passport-local';
+import { Request } from 'express';
+import { PassportStrategy } from '@nestjs/passport';
+import {
+  BadRequestException,
+  HttpException,
+  HttpStatus,
+  Injectable,
+  UnauthorizedException,
+  ValidationPipe,
+} from '@nestjs/common';
+import { User } from '../dtos/users/user.dto';
+import { PrismaService } from '../services/prisma.service';
+import { mapTo } from '../utilities/mapTo';
+import { defaultValidationPipeOptions } from '../utilities/default-validation-pipe-options';
+import { LoginViaSingleUseCode } from '../dtos/auth/login-single-use-code.dto';
+
+@Injectable()
+export class SingleUseCodeStrategy extends PassportStrategy(
+  Strategy,
+  'single-use-code',
+) {
+  constructor(private prisma: PrismaService) {
+    super({
+      usernameField: 'email',
+      passwordField: 'singleUseCode',
+      passReqToCallback: true,
+    });
+  }
+
+  /*
+    verifies that the incoming log in information is valid
+    returns the verified user
+  */
+  async validate(req: Request): Promise<User> {
+    const validationPipe = new ValidationPipe(defaultValidationPipeOptions);
+    const dto: LoginViaSingleUseCode = await validationPipe.transform(
+      req.body,
+      {
+        type: 'body',
+        metatype: LoginViaSingleUseCode,
+      },
+    );
+
+    if (!req?.headers?.jurisdictionname) {
+      throw new BadRequestException(
+        'jurisdictionname is missing from the request headers',
+      );
+    }
+
+    const jurisName = req.headers['jurisdictionname'];
+    const juris = await this.prisma.jurisdictions.findFirst({
+      where: {
+        name: {
+          in: Array.isArray(jurisName) ? jurisName : [jurisName],
+        },
+        allowSingleUseCodeLogin: true,
+      },
+    });
+    if (!juris) {
+      throw new BadRequestException(
+        'Single use code login is not setup for this jurisdiction',
+      );
+    }
```

Comment:
I think it would be useful to know if this error happens because the jurisdiction exists or if somebody passes in an invalid jurisdiction. Also it looks like we don't actually use the jurisdiction object. Would it make sense to fetch the jurisdiction using `findFirst` but just select the `allowSingleUseCodeLogin`. Then if `juris` doesn't exist that means the call was because it was a jurisdiction that doesn't exist and if `allowSingleUseCodeLogin` is not true we would throw the BadRequestException you have here.

Also as a side note, it would be worthwhile to log the jurisdiction name for these errors

Reply:
updated!